connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala

```diff
@@ -0,0 +1,191 @@
+     * has new progress to report.
+     *
+     * If the tracker was marked as dirty, the state is reset after.
+     */
+    def yieldWhenDirty(thunk: (Seq[StageInfo], Long) => Unit): Unit = {
+      if (dirty.get()) {
+        thunk(stages.values.toSeq, inFlightTasks.get())
+        dirty.set(false)
+      }
+    }
+
+    /**
+     * Add a job to the tracker. This will add the job to the list of jobs that are being tracked
+     */
+    def addJob(job: SparkListenerJobStart): Unit = synchronized {
+      jobs = jobs + job.jobId
+      job.stageInfos.foreach { stage =>
+        stages = stages + (stage.stageId -> new StageInfo(stage.stageId, stage.numTasks))
+      }
+      dirty.set(true)
+    }
+
+    def jobCount(): Int = {
+      jobs.size
+    }
+
+    def stageCount(): Int = {
+      stages.size
+    }
+  }
+
+  val trackedTags = collection.concurrent.TrieMap[String, ExecutionTracker]()
+
+  override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+    val tags = jobStart.properties.getProperty("spark.job.tags")
+    if (tags != null) {
+      val thisJobTags = tags.split(",").map(_.trim).toSet
+      thisJobTags.foreach { tag =>
+        trackedTags.get(tag).foreach { tracker =>
+          tracker.addJob(jobStart)
+        }
+      }
+    }
+  }
+
+  override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+    // Check if the task belongs to a job that we are tracking.
+    trackedTags.foreach({ case (_, tracker) =>
+      if (tracker.stages.contains(taskStart.stageId)) {
+        tracker.inFlightTasks.incrementAndGet()
+        tracker.dirty.set(true)
+      }
+    })
+  }
+
+  override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+    // Check if the task belongs to a job that we are tracking.
+    trackedTags.foreach({ case (_, tracker) =>
+      if (tracker.stages.contains(taskEnd.stageId)) {
+        tracker.stages.get(taskEnd.stageId).foreach { stage =>
+          stage.update { i =>
+            i.completedTasks += 1
+            i.inputBytesRead += taskEnd.taskMetrics.inputMetrics.bytesRead
+          }
+        }
+        // This should never become negative, simply reset to zero if it does.
+        tracker.inFlightTasks.decrementAndGet()
+        if (tracker.inFlightTasks.get() < 0) {
+          tracker.inFlightTasks.set(0)
+        }
+        tracker.dirty.set(true)
+      }
+    })
+  }
+
+  override def onStageCompleted(stageCompleted: SparkListenerStageCompleted): Unit = {
+    trackedTags.foreach({ case (_, tracker) =>
+      if (tracker.stages.contains(stageCompleted.stageInfo.stageId)) {
+        tracker.stages(stageCompleted.stageInfo.stageId).update { stage =>
+          stage.completed = true
+        }
+        tracker.dirty.set(true)
+      }
+    })
+  }
+
+  override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+    trackedTags.foreach({ case (_, tracker) =>
+      if (tracker.jobs.contains(jobEnd.jobId)) {
+        tracker.dirty.set(true)
```

Comment:
why do we set the dirty flag when nothing is updated?

Reply:
This is mostly to make sure that all progress is reported and an update is sent to the client. If you're tracking time between progress messages, every message itself is progress.